`ndarray.max()` does not call CUB in some conditions

As mentioned in #4232, a daily test fails where `ndarray.max()` does not call `cupy.core._routines_statistics.cub.device_segmented_reduce` for some reasons.

https://jenkins.preferred.jp/job/chainer/job/daily_master_cupy-shuffled/378/ID=20,label=mn1-p100/

The following shows the condition of the daily shuffle test.

```
11:22:17 + ./run_cupy_combination_test.py --id 20
11:22:17 --- Shuffle Parameters ---
11:22:17 cuda_libs: ('cuda90', 'cudnn73-cuda9', 'nccl2.4-cuda9', 'none')
11:22:17 scipy: None
11:22:17 base: ubuntu16_py35
11:22:17 numpy: 1.18
11:22:17 --- Configuration ---
11:22:17 nccl: nccl2.4-cuda9
11:22:17 cutensor: none
11:22:17 cudnn: cudnn73-cuda9
11:22:17 base: ubuntu16_py35
11:22:17 cuda: cuda90
11:22:17 requires: ['numpy<1.19']
11:23:55 sha256:28646ab8950d2c4793c87a7631006aa0a7c063fd6ba98cfeef674f1c0dff3798
```

Reproduced with python 3.6.
```
02:10:58 + ./run_cupy_combination_test.py --id 19
02:10:58 --- Shuffle Parameters ---
02:10:58 cuda_libs: ('cuda90', 'cudnn73-cuda9', 'nccl2.4-cuda9', 'none')
02:10:58 scipy: None
02:10:58 base: ubuntu16_py36-pyenv
02:10:58 numpy: 1.18
02:10:58 --- Configuration ---
02:10:58 nccl: nccl2.4-cuda9
02:10:58 cutensor: none
02:10:58 cudnn: cudnn73-cuda9
02:10:58 base: ubuntu16_py36-pyenv
02:10:58 cuda: cuda90
02:10:58 requires: ['numpy<1.19']
02:10:59 sha256:8fb2cfe5257855a59f5c616f846fdd44608eb96285007e73ba040c24c0f49715
```

<details>

<summary>tests.cupy_tests.core_tests.test_ndarray_reduction.TestCubReduction_param_0_{backend='device', order='C', shape=(10,)}.test_cub_max</summary>

### Error Message

```
AssertionError: Only cupy raises error    File "/work/cupy/cupy/testing/helper.py", line 37, in _call_func     result = impl(self, *args, **kw)   File "/work/cupy/tests/cupy_tests/core_tests/test_ndarray_reduction.py", line 409, in test_cub_max     a.max(axis=axis)   File "/opt/pyenv/versions/3.6.6/lib/python3.6/contextlib.py", line 88, in __exit__     next(self.gen)   File "/work/cupy/cupy/testing/helper.py", line 1446, in assert_function_is_called     assert handle.call_count == times_called
```

### Stacktrace

```
self = <<cupy.testing._bundle.TestCubReduction_param_0_{backend='device', order='C', shape=(10,)} testMethod=test_cub_max>  parameter: {'backend': 'device', 'order': 'C', 'shape': (10,)}>
xp = <module 'cupy' from '/work/cupy/cupy/__init__.py'>
dtype = <class 'numpy.float16'>, axis = (0,)

    @testing.for_contiguous_axes()
    @testing.for_all_dtypes(no_bool=True)
    @testing.numpy_cupy_allclose(rtol=1E-5)
    def test_cub_max(self, xp, dtype, axis):
        a = testing.shaped_random(self.shape, xp, dtype, order=self.order)
    
        if xp is numpy:
            return a.max(axis=axis)
    
        # xp is cupy, first ensure we really use CUB
        ret = cupy.empty(())  # Cython checks return type, need to fool it
        if self.backend == 'device':
            func_name = 'cupy.core._routines_statistics.cub.'
            if len(axis) == len(self.shape):
                func_name += 'device_reduce'
            else:
                func_name += 'device_segmented_reduce'
            with testing.AssertFunctionIsCalled(func_name, return_value=ret):
>               a.max(axis=axis)

a          = array([5.49 , 7.152, 6.027, 5.45 , 4.24 , 6.457, 4.375, 8.914, 9.63 ,
       3.834], dtype=float16)
axis       = (0,)
dtype      = <class 'numpy.float16'>
func_name  = 'cupy.core._routines_statistics.cub.device_reduce'
ret        = array(5.37910737e-299)
self       = <<cupy.testing._bundle.TestCubReduction_param_0_{backend='device', order='C', shape=(10,)} testMethod=test_cub_max>  parameter: {'backend': 'device', 'order': 'C', 'shape': (10,)}>
xp         = <module 'cupy' from '/work/cupy/cupy/__init__.py'>

tests/cupy_tests/core_tests/test_ndarray_reduction.py:409: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/pyenv/versions/3.6.6/lib/python3.6/contextlib.py:88: in __exit__
    next(self.gen)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

times_called = 1, args = ('cupy.core._routines_statistics.cub.device_reduce',)
kwargs = {'return_value': array(5.37910737e-299)}
handle = <MagicMock name='device_reduce' id='139983118485992'>

    @contextlib.contextmanager
    def assert_function_is_called(*args, times_called=1, **kwargs):
        """A handy wrapper for unittest.mock to check if a function is called.
    
        Args:
            *args: Arguments of `mock.patch`.
            times_called (int): The number of times the function should be
                called. Default is ``1``.
            **kwargs: Keyword arguments of `mock.patch`.
    
        """
        with mock.patch(*args, **kwargs) as handle:
            yield
>           assert handle.call_count == times_called
E           AssertionError: Only cupy raises error
E           
E             File "/work/cupy/cupy/testing/helper.py", line 37, in _call_func
E               result = impl(self, *args, **kw)
E             File "/work/cupy/tests/cupy_tests/core_tests/test_ndarray_reduction.py", line 409, in test_cub_max
E               a.max(axis=axis)
E             File "/opt/pyenv/versions/3.6.6/lib/python3.6/contextlib.py", line 88, in __exit__
E               next(self.gen)
E             File "/work/cupy/cupy/testing/helper.py", line 1446, in assert_function_is_called
E               assert handle.call_count == times_called

args       = ('cupy.core._routines_statistics.cub.device_reduce',)
handle     = <MagicMock name='device_reduce' id='139983118485992'>
kwargs     = {'return_value': array(5.37910737e-299)}
times_called = 1

cupy/testing/helper.py:1446: AssertionError
```

### Standard Output

```
dtype is <class 'numpy.float16'>
axis is (0,) , ndim is 1 , shape is (10,) , order is C
```

</details>

https://jenkins.preferred.jp/job/chainer/job/daily_master_cupy-shuffled/388/ID=19,label=mn1-p100/testReport/junit/tests.cupy_tests.core_tests.test_ndarray_reduction/TestCubReduction_param_0_%7Bbackend='device',%20order='C',%20shape=(10,)%7D/test_cub_max/

The error message in another failing test is a lot clearer:
```
>   raise RuntimeError('either the GPU or the CUDA Toolkit does not '
E   AssertionError: Only cupy raises error
E   
E     File "/work/cupy/cupy/testing/helper.py", line 37, in _call_func
E       result = impl(self, *args, **kw)
E     File "/work/cupy/tests/cupy_tests/logic_tests/test_truth.py", line 117, in test
E       return getattr(xp, self.f)(x, y, self.assume_unique, self.invert)
E     File "/work/cupy/cupy/_logic/truth.py", line 95, in in1d
E       ar2 = cupy.sort(ar2)
E     File "/work/cupy/cupy/_sorting/sort.py", line 32, in sort
E       ret.sort(axis=axis)
E     File "cupy/core/core.pyx", line 636, in cupy.core.core.ndarray.sort
E       cpdef sort(self, int axis=-1):
E     File "cupy/core/core.pyx", line 654, in cupy.core.core.ndarray.sort
E       _sorting._ndarray_sort(self, axis)
E     File "cupy/core/_routines_sorting.pyx", line 45, in cupy.core._routines_sorting._ndarray_sort
E       thrust.sort(self.dtype, data.data.ptr, 0, self.shape)
E     File "cupy/cuda/thrust.pyx", line 75, in cupy.cuda.thrust.sort
E       cpdef sort(dtype, intptr_t data_start, intptr_t keys_start,
E     File "cupy/cuda/thrust.pyx", line 90, in cupy.cuda.thrust.sort
E       raise RuntimeError('either the GPU or the CUDA Toolkit does not '
E   either the GPU or the CUDA Toolkit does not support fp16
```
https://jenkins.preferred.jp/job/chainer/job/daily_master_cupy-shuffled/388/ID=19,label=mn1-p100/testReport/junit/tests.cupy_tests.logic_tests.test_truth/TestIn1DIsIn_param_7_%7Bassume_unique=False,%20f='in1d',%20invert=False,%20shape_x=(3,),%20shape_y=(3,)%7D/test/

--- a/tests/cupy_tests/core_tests/test_ndarray_reduction.py
+++ b/tests/cupy_tests/core_tests/test_ndarray_reduction.py
@@ -324,6 +324,13 @@ class TestArrayReduction(unittest.TestCase):
         return a.argmin()
 
 
+# TODO(leofang): remove this once CUDA 9.0 is dropped
+def _skip_cuda90(dtype):
+    ver = cupy.cuda.runtime.runtimeGetVersion()
+    if dtype == cupy.float16 and ver == 9000:
+        pytest.skip('CUB does not support fp16 on CUDA 9.0')
+
+
 # This class compares CUB results against NumPy's
 @testing.parameterize(*testing.product({
     'shape': [(10,), (10, 20), (10, 20, 30), (10, 20, 30, 40)],
@@ -352,6 +359,7 @@ class TestCubReduction(unittest.TestCase):
     @testing.for_all_dtypes(no_bool=True)
     @testing.numpy_cupy_allclose(rtol=1E-5)
     def test_cub_min(self, xp, dtype, axis):
+        _skip_cuda90(dtype)
         a = testing.shaped_random(self.shape, xp, dtype, order=self.order)
 
         if xp is numpy:
@@ -392,6 +400,7 @@ class TestCubReduction(unittest.TestCase):
     @testing.for_all_dtypes(no_bool=True)
     @testing.numpy_cupy_allclose(rtol=1E-5)
     def test_cub_max(self, xp, dtype, axis):
+        _skip_cuda90(dtype)
         a = testing.shaped_random(self.shape, xp, dtype, order=self.order)
 
         if xp is numpy:

--- a/tests/cupy_tests/sorting_tests/test_search.py
+++ b/tests/cupy_tests/sorting_tests/test_search.py
@@ -161,6 +161,13 @@ class TestSearch(unittest.TestCase):
         return a.argmin(axis=1)
 
 
+# TODO(leofang): remove this once CUDA 9.0 is dropped
+def _skip_cuda90(dtype):
+    ver = cupy.cuda.runtime.runtimeGetVersion()
+    if dtype == cupy.float16 and ver == 9000:
+        pytest.skip('CUB does not support fp16 on CUDA 9.0')
+
+
 # This class compares CUB results against NumPy's
 # TODO(leofang): test axis after support is added
 @testing.parameterize(*testing.product({
@@ -192,6 +199,7 @@ class TestCubReduction(unittest.TestCase):
     @testing.for_dtypes('bhilBHILefdFD')
     @testing.numpy_cupy_allclose(rtol=1E-5, contiguous_check=False)
     def test_cub_argmin(self, xp, dtype):
+        _skip_cuda90(dtype)
         a = testing.shaped_random(self.shape, xp, dtype)
         if self.order == 'C':
             a = xp.ascontiguousarray(a)
@@ -226,6 +234,7 @@ class TestCubReduction(unittest.TestCase):
     @testing.for_dtypes('bhilBHILefdFD')
     @testing.numpy_cupy_allclose(rtol=1E-5, contiguous_check=False)
     def test_cub_argmax(self, xp, dtype):
+        _skip_cuda90(dtype)
         a = testing.shaped_random(self.shape, xp, dtype)
         if self.order == 'C':
             a = xp.ascontiguousarray(a)

